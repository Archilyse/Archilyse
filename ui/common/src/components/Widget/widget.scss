@import '../../theme.scss';

.widget-drawer {
  z-index: $z-index-level-1;
  background-color: $white;
  position: relative;
  right: 0;
  transition: width 0.3s;

  &.collapsed {
    & * {
      max-width: 80px;
    }
  }
}

.widget-expanded-space {
  transition: width 0.3s;
}

.tabs {
  display: flex;
  flex-direction: column;
  height: 100%;

  .tab-headers {
    display: flex;
    flex: 0 1 $unit * 5;
    min-height: $unit * 5;
    width: fit-content;

    .tab {
      display: flex;
      align-items: center;
      padding: 0 $unit * 2;
      border-bottom: 2px solid transparent;
      color: $medium-gray;
      font-weight: 600;
      cursor: pointer;

      .name {
        flex: 1;
        overflow: hidden;
        white-space: nowrap;
        text-overflow: ellipsis;
        text-align: center;
      }

      .dots {
        margin-left: auto;
      }

      &:hover {
        color: $outer-space-grey;
      }
    }

    .tab.selected {
      z-index: 10;
      border-bottom: 2px solid $outer-space-grey;
      color: $outer-space-grey;
      font-weight: 600;
    }
  }

  .tab-content {
    flex: 1 1 auto;
    border-top: 1px solid $bright-gray;

    &.collapsed {
      position: absolute;
      top: -9999px;
      left: -9999px;
    }
  }
}

.collapse-widget-button {
  display: inline-flex;
  justify-content: center;
  align-items: center;

  border: none;
  color: $primary-color;
  width: $unit * 2;
  height: $unit * 2;

  margin-left: 4px;
  margin-right: 4px;
  outline: none;

  p {
    margin: 0;
    position: relative;
    bottom: 2px;
    font-weight: 400;
  }
}
