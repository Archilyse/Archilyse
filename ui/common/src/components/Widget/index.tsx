import React, { useEffect, useRef, useState } from 'react';
import classnames from 'classnames';
import { Tooltip } from '@material-ui/core';
import './widget.scss';

const SLIDE_OFFSET = 80;

const CollapseButton = ({ collapsed, onClick }) => {
  const handleClick = event => {
    // prevent from changing a tab together with toggling collapse state
    event.stopPropagation();

    onClick();
  };

  return (
    <Tooltip title={collapsed ? 'Expand' : 'Collapse'}>
      <div className="collapse-widget-button" onClick={handleClick}>
        <p>{collapsed ? '<' : '>'}</p>
      </div>
    </Tooltip>
  );
};

const Tabs = ({ headers, children, onCollapse, collapsed, tab = 0, onChangeTab, extraContent }) => {
  const selectedChild = children.find((component, index) => (collapsed ? index === 0 : index === tab));

  const isTabSelected = (tab, index) => (collapsed ? index === 0 : tab === index);

  const tabs = collapsed ? headers.slice(0, 1) : headers;
  return (
    <div className="tabs">
      <div className="tab-headers">
        {tabs.map((tabName, index) => (
          <div
            key={tabName}
            className={`tab ${isTabSelected(tab, index) ? 'selected' : ''}`}
            onClick={() => onChangeTab(index)}
          >
            {isTabSelected(tab, index) && onCollapse && <CollapseButton onClick={onCollapse} collapsed={collapsed} />}
            <div className="name">{tabName}</div>
            {extraContent[index]}
          </div>
        ))}
      </div>
      <div className={classnames('tab-content', { collapsed })}>{selectedChild}</div>
    </div>
  );
};

const getStyle = (widget, fixedWidth, collapsed) => {
  if (!widget) return { width: fixedWidth };

  const { right } = widget.getBoundingClientRect();
  const width = collapsed ? SLIDE_OFFSET : fixedWidth;
  const x = collapsed ? window.innerWidth - right : 0;

  return { width, transform: `translateX(${x}px)` };
};

const Widget = ({
  collapsed: defaultCollapsed = false,
  onCollapse = null,
  onTabChange = null,
  className = '',
  children,
  tabHeaders,
  initialTab = 0,
  width = '100%',
  extraContent = [],
}) => {
  const [tab, setTab] = useState(initialTab);
  const [collapsed, setCollapsed] = useState(defaultCollapsed);
  const [style, setStyle] = useState({ width });
  const widget = useRef();

  const handleChangeTab = (index: number) => {
    setTab(index);

    if (onTabChange) onTabChange(index);
  };

  const handleCollapse = () => {
    setStyle(getStyle(widget.current, width, !collapsed));
    setCollapsed(!collapsed);
    onCollapse();
  };

  useEffect(() => {
    setTab(initialTab);
  }, [initialTab]);

  useEffect(() => {
    setCollapsed(defaultCollapsed);
  }, [defaultCollapsed]);

  const filteredChildren = React.Children.toArray(children);
  const classes = classnames('widget-drawer', className, { collapsed });

  return (
    <div data-testid="widget" ref={widget} style={style} className={classes}>
      <Tabs
        headers={tabHeaders}
        tab={tab}
        onCollapse={onCollapse ? handleCollapse : null}
        onChangeTab={handleChangeTab}
        collapsed={collapsed}
        extraContent={extraContent}
      >
        {filteredChildren}
      </Tabs>
    </div>
  );
};

export default Widget;
