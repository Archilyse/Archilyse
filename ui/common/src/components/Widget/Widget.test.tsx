import * as React from 'react';
import { cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import Widget from './index';

afterEach(cleanup);

const MOCK_CONTENT_1 = 'Wonderful tab content 1';
const MOCK_CONTENT_2 = 'Wonderful tab content 2';

it('renders correctly with two tabs', () => {
  const { container } = render(
    <Widget className={'view-drawer'} tabHeaders={['Tab 1', 'Tab 2']}>
      <p>{MOCK_CONTENT_1}</p>
      <p>{MOCK_CONTENT_2}</p>
    </Widget>
  );
  expect(container).toMatchSnapshot();
});

it('change tabs successfully', async () => {
  const { getByText } = render(
    <Widget className={'view-drawer'} tabHeaders={['Tab 1', 'Tab 2']}>
      <p>{MOCK_CONTENT_1}</p>
      <p>{MOCK_CONTENT_2}</p>
    </Widget>
  );
  const isTabVisible = tabContent => getByText(tabContent);

  expect(isTabVisible(MOCK_CONTENT_1)).toBeTruthy();

  // Switch tab and ensure the content has changed
  fireEvent.click(getByText('Tab 2'));

  await waitFor(() => expect(isTabVisible(MOCK_CONTENT_2)).toBeTruthy());
  expect(isTabVisible(MOCK_CONTENT_2)).toBeTruthy();
});
