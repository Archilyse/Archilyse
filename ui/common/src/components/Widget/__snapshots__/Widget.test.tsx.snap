// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with two tabs 1`] = `
<div>
  <div
    class="widget-drawer view-drawer"
    data-testid="widget"
    style="width: 100%;"
  >
    <div
      class="tabs"
    >
      <div
        class="tab-headers"
      >
        <div
          class="tab selected"
        >
          <div
            class="name"
          >
            Tab 1
          </div>
        </div>
        <div
          class="tab "
        >
          <div
            class="name"
          >
            Tab 2
          </div>
        </div>
      </div>
      <div
        class="tab-content"
      >
        <p>
          Wonderful tab content 1
        </p>
      </div>
    </div>
  </div>
</div>
`;
